reco_utils/common/python_utils.py

```diff
@@ -2,29 +2,50 @@
 from scipy.sparse import coo_matrix
 
 
+def exponential_decay(value, max_val, half_life):
+    """Compute decay factor for a given value based on an exponential decay
+    Args:
+        value (numeric): value to calculate decay factor
+        max_val (numeric): value at which decay factor will be 1
+        half_life (numeric): value at which decay factor will be 0.5
+    Returns:
+        (float): decay factor
+    """
+
+    return np.exp(-np.log(2) * (max_val - value) / half_life)
```

Comment:
I have a nice piece of code for you. Can you try this and see if it is faster on SAR?

```
def exponential_decay(value, max_val, half_life):
    return np.exp(-np.log(2) * (max_val - value) / half_life)

def exponential_decay_ne(value, max_val, half_life):
    return ne.evaluate("exp(-log(2)*(max_val - value) / half_life)")

s = 1000000
aa = np.random.randn(s, 1).astype(np.float32)
e1 = exponential_decay(aa, 5, 2)
e2 = exponential_decay_ne(aa, 5, 2)
np.allclose(e1, e2) #True
%timeit exponential_decay(aa, 5, 2) #11.2 ms ± 22.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit exponential_decay_ne(aa, 5, 2) #2.27 ms ± 57.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
👍

Reply:
is it faster?

Reply:
for ML-1M
numexpr takes: 11.911 ms
numpy takes: 75.956 ms

I wish they would just merge numexpr with numpy!

Reply:
awesome!!!! do you mind to get the metrics for 10M and 20M?

Reply:
| Dataset | Impl | Time (ms)|
|-|-|-|
| 1M | numexpr | 47.705 |
| 1M | numpy | 19.273 |
| 10M | numexpr | 264.781 |
| 10M | numpy | 138.637 |
| 20M | numexpr | 383.995 |
| 20M | numpy | 235.256 |

Oops, I think I posted the last results backwards, sorry!

ugh, at this point though it seems like I should revert to the original numpy code

Reply:
mmm, this is interesting, I wonder why this is the behavior. numexpr should be faster than numpy

Reply:
i wonder if it has something to do with how the data is being passed in, since we're passing in a pandas series numpy should be able to vectorize the computation, maybe numexpr is looping over each value?

Reply:
so i looked at evaluating those two expressions on synthetic data, numexpr has an initial cost to compile / start up which means it needs a larger data size to out-perform numpy. 

This [thread](https://github.com/pydata/numexpr/issues/301#issuecomment-388097698) seems to indicate that size is in the 100k range, but from my test it's closer to the 100M range, which we're not reaching with these datasets.

I can send the evaluation code if you want to duplicate it (for some reason it won't let me attach the file here), not sure if the versions of numpy and numexpr have anything to do with it Maybe optimizations happening that were not there in older versions of numpy?

![numpy_vs_numexpr](https://user-images.githubusercontent.com/5720537/52450467-18a9ec00-2b09-11e9-8f3c-3ffe26889159.jpg)